### PR TITLE
[Merged by Bors] - feat: coalgebra homomorphisms

### PR DESCRIPTION
**Pull request by 101damnations**
Homomorphisms of coalgebras, mimicking the file `Mathlib.Algebra.Algebra.Hom`.

---

I've left a few comments in the code on things I'm unsure of the purpose of. 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by kbuzzard — APPROVED** [review]
This all looks very uncontroversial to me. As you say, it's following Algebra.Algebra.Hom .

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  [AddCommMonoid A] [Module R A] [AddCommMonoid B] [Module R B]
+  [CoalgebraStruct R A] [CoalgebraStruct R B] [FunLike F A B]
+  [CoalgHomClass F R A B]
+
+/-- Turn an element of a type `F` satisfying `CoalgHomClass F R A B` into an actual
+`CoalgHom`. This is declared as the default coercion from `F` to `A →ₗc[R] B`. -/
+@[coe]
+def toCoalgHom (f : F) : A →ₗc[R] B :=
+  { (f : A →ₗ[R] B) with
+    toFun := f -- is this necessary? Emulating `AlgHomClass.toAlgHom`
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:69]: I should at some point try to revive #6791; new-style structures are just a waste of everyone's time for morphisms, where we always take everything apart and put it back together anyway.

[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:69]: Oh interesting! I hadn't realised that. So for morphisms the way we pack things up doesn't matter like it does for classes with data.

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:69]: Basically no; the only thing we really do with morphisms is unfold them completely to their `toFun` field, at which point it doesn't matter how we put them together.

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+@[simp]
+theorem toLinearMap_eq_coe (f : A →ₗc[R] B) : f.toLinearMap = f :=
+  rfl
+
+@[simp, norm_cast]
+theorem coe_toLinearMap (f : A →ₗc[R] B) : ⇑(f : A →ₗ[R] B) = f :=
+  rfl
+
+-- simp can prove this
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Coalgebra/Hom.lean:154]: Can it? Then why isn't there a lint error?

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+variable [CommSemiring R] [AddCommGroup A] [AddCommGroup B] [Module R A] [Module R B]
+
+variable [CoalgebraStruct R A] [CoalgebraStruct R B] (φ : A →ₗc[R] B)
+
+protected theorem map_neg (x) : φ (-x) = -φ x :=
+  map_neg _ _
+
+protected theorem map_sub (x y) : φ (x - y) = φ x - φ y :=
+  map_sub _ _ _
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Coalgebra/Hom.lean:269]: Do you need these?

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+theorem counitCoalgHom_apply (x : A) :
+    counitCoalgHom R A x = counit x := rfl
+
+@[simp]
+theorem counitCoalgHom_toLinearMap :
+    counitCoalgHom R A = counit (R := R) (A := A) := rfl
+
+variable {R}
+
+instance subsingleton_to_ring : Subsingleton (A →ₗc[R] R) :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Coalgebra/Hom.lean:300]: What does `to_ring` mean here (and below)?

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+@[simp]
+protected theorem coe_coe {F : Type*} [FunLike F A B] [CoalgHomClass F R A B] (f : F) :
+    ⇑(f : A →ₗc[R] B) = f :=
+  rfl
+
+-- removed @[simp], linter complains
+theorem toFun_eq_coe (f : A →ₗc[R] B) : f.toFun = f := rfl
+
+-- are the next 2 declarations necessary? Can already coerce to an `AddMonoidHom`
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Coalgebra/Hom.lean:126]: Are they?

File: Mathlib/RingTheory/Coalgebra/Hom.lean [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+initialize_simps_projections CoalgHom (toFun → apply)
+
+@[simp]
+protected theorem coe_coe {F : Type*} [FunLike F A B] [CoalgHomClass F R A B] (f : F) :
+    ⇑(f : A →ₗc[R] B) = f :=
+  rfl
+
+-- removed @[simp], linter complains
+theorem toFun_eq_coe (f : A →ₗc[R] B) : f.toFun = f := rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Coalgebra/Hom.lean:124]: Complains why? Do you even need it?

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 2 above)
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:154]: Yep. I don't know why there isn't a lint error. I'll remove the `@[simp]` anyway.

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 3 above)
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:269]: Nope. There are loads of similar `protected` theorems in `Algebra.Algebra.Hom` I chose not to emulate and I didn't intend to keep these ones in particular.

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 4 above)
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:300]: I think they were inspired by [`LinearMap.ext_ring`](https://github.com/leanprover-community/mathlib4/blob/e8433a6b8e7980bfaaf9b34513b77da2af14995a/Mathlib/Algebra/Module/LinearMap/Basic.lean#L497). Maybe `ext_counit` or something is better?

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 5 above)
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:126]: Not as far as I can tell. Do you know why [these](https://github.com/leanprover-community/mathlib4/blob/e8433a6b8e7980bfaaf9b34513b77da2af14995a/Mathlib/Algebra/Algebra/Hom.lean#L134-L148) exist in `Algebra.Algebra.Hom`?

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 6 above)
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:124]: LHS simplifies to `⇑f.toLinearMap`; indeed `simp` can prove it if I move it a few lines later (where the linter also doesn't complain if I tag it with `@[simp]`... I must've misunderstood simp linting). I'll remove this lemma.

File: Mathlib/RingTheory/Coalgebra/Hom.lean (hunk 4 above)
Comment (kbuzzard) [reply Mathlib/RingTheory/Coalgebra/Hom.lean:274]: I think the current name is easier to find than `ext_counit`.

**Comment by kbuzzard** [comment]
Thanks a lot!

bors merge

**Pull request closed without merging**